src/dev/precommit_hook/check_lockfile_symlinks.js

```diff
@@ -0,0 +1,149 @@
+
+import { dirname } from 'path';
+import { existsSync, lstatSync, readFileSync } from 'fs';
+
+import { createFailError } from '@kbn/dev-utils';
+import { matchesAnyGlob } from '../globs';
+
+import { LOCKFILE_GLOBS, MANIFEST_GLOBS, IGNORE_FILE_GLOBS } from './lockfile_check_config';
+
+function listPaths(paths) {
+  return paths.map(path => ` - ${path}`).join('\n');
+}
+
+async function checkProhibitedLockfiles(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink() && parent !== '.') {
+        errorPaths.push(path);
+      }
+    });
+
+  if (errorPaths.length) {
+    throw createFailError(
+      `These directories MUST NOT have a 'yarn.lock' file:\n${listPaths(errorPaths)}`
+    );
+  }
+}
+
+async function checkSuperfluousSymlinks(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink()) {
+        return;
+      }
+
+      const manifestPath = `${parent}/package.json`;
+      if (!existsSync(manifestPath)) {
+        log.warning(
+          `No manifest found at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+        );
+        errorPaths.push(path);
+        return;
+      }
+
+      try {
+        const manifest = readFileSync(manifestPath);
+        try {
+          const json = JSON.parse(manifest);
+          if (!json.dependencies || !Object.keys(json.dependencies).length) {
+            log.warning(
+              `Manifest at '${manifestPath}' has no dependencies, but found an adjacent 'yarn.lock' symlink.`
+            );
+            errorPaths.push(path);
+          }
+        } catch (err) {
+          log.warning(
+            `Could not parse JSON at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+          );
+          errorPaths.push(path);
+        }
+      } catch (err) {
+        log.warning(
+          `Could not read manifest at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+        );
+        errorPaths.push(path);
+      }
+    });
+
+  if (errorPaths.length) {
+    throw createFailError(
+      `These directories MUST NOT have a 'yarn.lock' symlink:\n${listPaths(errorPaths)}`
+    );
+  }
+}
+
+async function checkMissingSymlinks(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), MANIFEST_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const lockfilePath = `${parent}/yarn.lock`;
+      if (existsSync(lockfilePath)) {
+        return;
+      }
```

Comment:
We don't need to check if the file is a symlink, because if it's _not_, it will throw an error during `checkProhibitedLockfiles`. So, we assume if a lockfile exists here, it's a symlink and we don't need to proceed.

I thought about using `fs.realpathSync` to check and ensure that the symlink is pointing to the appropriate `yarn.lock` file in the project root directory, but it seemed unnecessary. Or, if we decide it is necessary, maybe we should put it in a separate method to ensure that the symlinks that exist are set appropriately.